### ISSUE TITLE
Document that entity data must be consumed both for requests and responses to avoid the connection to stall

### ISSUE THREAD

**Issue by samueltardieu** [akka/akka#17591]
The context: I’m using an in-house library to access CouchDB databases over HTTP named *canape*. I create a database and wait for the result, I then open a long-running connection to the *_changes* stream which returns live database changes and send them to a *fold* sink in the background, and I immediately create 5 documents named *docid* and wait for 100 milliseconds after every document creation.

In the example below, we have a single TCP stream dumped with wireshark. Note how the request to *_changes* (which represents a live stream of changes in CouchDB) returns a chunked response, and the chunked response is obviously not over when the `PUT` to create the second document (*docid2*) is sent over the same connection. Even though I took care of using `POST` to ensure that Akka does see the *_changes* request as non-idempotent.

Note that the first document (*docid1*) has been created on another TCP connection (and thus not shown here), probably because it has been sent right after the request to the *_changes* stream, which means that at this time Akka probably considered the non-idempotent request to be running. However, the second document is being created on the busy original TCP connection, as if as soon as the headers were sent back the connection is considered idle again, although its entity is still being transmitted and may be for a long time. Since the `PUT` for *docid2* is still blocked on this connection, we can also see that *docid3*, *docid4* and *docid5* have been properly created using another TCP connection.

Right now, I work around this by creating a new host connection pool for every long-running connection, but it is a waste or resources and causes some code duplication. The problem shown below happens when I try to use a common pool for all requests (which should work fine).

```
> PUT /canape-test-db-d45ad182-e2ec-4387-b5a1-a51d016c0312/ HTTP/1.1 
> Host: localhost:5984 
> User-Agent: canape for Scala 
> Accept: application/json 
> Content-Length: 0 
>  
< HTTP/1.1 201 Created 
< Server: CouchDB/1.6.1 (Erlang OTP/17) 
< Location: http://localhost:5984/canape-test-db-d45ad182-e2ec-4387-b5a1-a51d016c0312 
< Date: Wed, 27 May 2015 16:43:24 GMT 
< Content-Type: application/json 
< Content-Length: 12 
< Cache-Control: must-revalidate 
<  
< {"ok":true}
> POST /canape-test-db-d45ad182-e2ec-4387-b5a1-a51d016c0312/_changes?feed=continuous HTTP/1.1 
> Host: localhost:5984 
> User-Agent: canape for Scala 
> Accept: application/json 
> Content-Type: application/json 
> Content-Length: 2 
>  
> {}
< HTTP/1.1 200 OK 
< Transfer-Encoding: chunked 
< Server: CouchDB/1.6.1 (Erlang OTP/17) 
< Date: Wed, 27 May 2015 16:43:25 GMT 
< Content-Type: application/json 
< Cache-Control: must-revalidate 
<  
< 51 
< {"seq":1,"id":"docid1","changes":[{"rev":"1-967a00dff5e02add41819138abb3284d"}]}
<  
> PUT /canape-test-db-d45ad182-e2ec-4387-b5a1-a51d016c0312/docid2 HTTP/1.1 
> Host: localhost:5984 
> User-Agent: canape for Scala 
> Accept: application/json 
> Content-Type: application/json 
> Content-Length: 2 
>  
> {}
< 51 
< {"seq":2,"id":"docid3","changes":[{"rev":"1-967a00dff5e02add41819138abb3284d"}]}
<  
< 51 
< {"seq":3,"id":"docid4","changes":[{"rev":"1-967a00dff5e02add41819138abb3284d"}]}
<  
< 51 
< {"seq":4,"id":"docid5","changes":[{"rev":"1-967a00dff5e02add41819138abb3284d"}]}
```

(hangs there, as the connection is still busy sending the long-running _changes stream, the creation of *docid2* will timeout)

**Comment by rkuhn** [comment]
cc @sirthias

**Comment by rkuhn** [comment]
This is also a general problem for long-running GET requests that are considered idempotent, but perhaps we need another ticket for that—this ticket is about the connection being considered idle prematurely.

**Comment by sirthias** [comment]
This appears to be a duplicate of #17403, no?

**Comment by samueltardieu** [comment]
@sirthias I don't think so, as the "_changes" request is still executing and its entity is being read. If I put a `println` in the fold, I clearly see the creation of *docid1*, *docid3*, *docid4* and *docid5*, which means that the source associated to the slot is still being drained after the pipelined request for creating *docid2* has been sent onto the same busy slot.

Here is what the test looks like:

```scala
  "db.continuousChanges()" should {

    "see the creation of new documents" in new freshDb {
      implicit val materializer = ActorFlowMaterializer(None)
      val changes: Source[JsObject, Unit] = db.continuousChanges()
      val result = changes.via(Database.onlySeq).map(j => (j \ "id").as[String]).take(5).runFold[List[String]](Nil) {
        case (l, e) =>
          println(s"Received $e")
          l :+ e
      }
      for (i <- 1 to 5) {
        waitEventually(db.insert(JsObject(Nil), s"docid$i"))
        Thread.sleep(100)
      }
      waitForResult(result).sorted must be equalTo List("docid1", "docid2", "docid3", "docid4", "docid5")
    }

  }
```

During the execution it prints:

```
Received docid1
Received docid3
Received docid4
Received docid5
```

which is consistent with the wireshark trace (*docid2* being queued onto the same slot which is currently receiving this information).

Some explanations: `onlySeq` filters lines containing a JSON object with a `seq` field, which means all the lines sent by the "_changes" feed so far, `waitEventually` remembers the `Future` returned by `db.insert` to perform some cleanup at the end of the test, and `waitForResult` is `Await.result` with a fixed timeout.

Also, the `take(5)` is here to make the test end sooner eventually but should not cause any change of behaviour since we wait for 5 objects and we receive only 4. And removing it does not change anything. So I am pretty sure it is not similar to #17403 as the slot data is obviously still being consumed.

**Comment by rkuhn** [comment]
There are multiple angles here:

* not consuming a reply payload should drain it from the connection (which I understand #17403 to be about)
* connections are put back into the pool after being blocked (for non-idempotent requests) too early, namely before the payload is consumed for the last response (which is this ticket)
* we have no way of marking a request as “long-polling” to disable further pipelining on that connection even for idempotent requests (which does not yet have a ticket, I think)

**Comment by jrudolph** [comment]
> we have no way of marking a request as “long-polling” to disable further pipelining on that connection even for idempotent requests (which does not yet have a ticket, I think)

Just use a pool with pipelining disabled (which is the default). It looks as if the issue here is that things are pipelined even without pipelining being requested. Or did you change the pipelining setting, @samueltardieu?

**Comment by samueltardieu** [comment]
Btw, that makes me think that I also suspected it to be similar to #17403 and that I originally inserted a flow which drains the source when the sink unsubscribes before the `take(5)` (using an ugly flexiroute hack), and it didn't change anything for my problem.

**Comment by samueltardieu** [comment]
@jrudolph I didn't. And I even tried with a pool with pipelining explicitly disabled (pipeliningLimit = 1) because I wasn't sure of the default. Which reinforces my feeling that the slot might be considered idle too soon as far as slot determination is concerned, as @rkuhn summarizes above in his second bullet item.

**Comment by rkuhn** [comment]
Yes, I do think that we have three distinct problems here, no doubt about it.

**Comment by jrudolph** [comment]
Then we have
* a bug in scheduling requests too early on non-pipelining connections (this ticket and #17403)
* an API inconvenience that you need to consume entity data (the thing that triggered #17403)

What's the third one?

**Comment by rkuhn** [comment]
The third is that it is currently not possible to use shared pipelining connection pools for long-polling GET requests. We might decide that that is how it is, or we might decide to add something to the `HttpRequest` that informs the pool of the request’s semantics, but in any case we’ll have to document it.

**Comment by samueltardieu** [comment]
I opened #17595 for this third one.

**Comment by sirthias** [comment]
Gentlemen, I see only one issue here, namely #17403.
IMHO this ticket here is a duplicate.

The "API inconvenience that you need to consume entity data" is not a bug, it's a necessary consequence of the streaming nature of Akka HTTP and needs to be addressed through documentation.
The problem that it is not possible to use shared pipelining connection pools for long-polling GET requests is not a bug IMO.

**Comment by rkuhn** [comment]
Well, then we have one bug and two documentation issues :-)

**Comment by sirthias** [comment]
@rkuhn Ok, let's settle on that. :)
I'll submit a fix to the bug shortly.

**Comment by jrudolph** [comment]
This original issue (to make sure to consume entities) seems to be already documented in a "Caution" box on this page: http://doc.akka.io/docs/akka-stream-and-http-experimental/1.0/scala/http/common/http-model.html#HttpEntity

I added a commit to #18047 that adds a note to the host-level and request-level API documentation not to use them for long-running requests.

**Comment by francisdb** [comment]
Is there no way to warn the user about this at runtime? Preferably a logged warning.

All I see in the logs is:
```
root 17:17:33.994 DEBUG   akka.stream.impl.PrefixAndTailImpl - Cancelling akka.stream.impl.MultiStreamOutputProcessor$SubstreamOutput@52d14d67 (after: 5000 ms)
root 17:17:34.003 DEBUG   akka.stream.impl.PrefixAndTailImpl - Cancelling akka.stream.impl.MultiStreamOutputProcessor$SubstreamOutput@6528aaae (after: 5000 ms)
```
and no new responses coming back after 4 requests

Searching on this will now hopefully point users here!
 
What is the fastest way to consume/drop the response entity?

**Comment by ktoso** [comment]
Fastest is most likely to pipe it to a `Sink.ignore`

**Comment by francisdb** [comment]
hmm, how to pipe HttpResponse.entity?

**Comment by jrudolph** [comment]
```scala
response.entity.dataBytes.runTo(Sink.ignore)
```
